File: [1223][Medium][Dice_Roll_Simulation]/Dice_Roll_Simulation.py
# leetcode time     cost : 2192 ms 
# leetcode memory   cost : 21.9 MB 
# solution 1, DP
class Solution:
    def dieSimulator(self, n: int, rollMax: [int]) -> int:
        dp = [[[0 for _ in range(16)] for _ in range(7)] for _ in range(n + 1)]
        mod = 10**9 + 7
                
        for i in range(1, n + 1):
            # 投掷的数
            for j in range(1, 7):
                # 第一次投掷
                if i == 1:
                    dp[i][j][1] = 1
                    continue
                
                # 数字 j 连续出现 k 次
                for k in range(2, rollMax[j - 1] + 1):
                    dp[i][j][k] = dp[i - 1][j][k - 1]
                    
                # 前一次投出的数不是 j
                s = 0
                for l in range(1, 7):
                    if l == j:
                        continue
                    for k in range(1, 16):
                        s += dp[i - 1][l][k]
                        s %= mod
                dp[i][j][1] = s
        
        res = 0
        for j in range(1, 7):
            for k in range(1, 16):
                # 求投掷 n 次时所有组合总和
                res += dp[n][j][k]
                res %= mod
                
        return res
    
def main():
    n, rollMax = 2,[1,1,2,2,2,3]         # expect is 34
    obj = Solution()
    result = obj.dieSimulator(n, rollMax)
    print("return result is :",result)
    
if __name__ =='__main__':
    main()

File: [1223][Medium][Dice_Roll_Simulation]/Dice_Roll_Simulation.cpp
// leetcode time     cost : 304 ms
// leetcode memory   cost : 34.3 MB 
// Time  Complexity: O(N*J*K)
// Space Complexity: O(N*J*K)
// solution 1, DP, 
#include <vector>
using namespace std;
class Solution {
public:
    int dieSimulator(int n, vector<int>& rollMax) {
        const int divider = 1e9 + 7;
        vector<vector<vector<int>>> dp(n,vector<vector<int>>(7,vector<int>(16,0)));//与用三维数组同理
        for (int i = 1; i < 7; ++i) {
            dp[0][i][1] = 1;
        }
        for (int i = 1; i < n; ++i) {//第i + 1次投掷
            for (int j = 1; j < 7; ++j) {//出现的点数为j
                for (int k = 1; k < 7; ++k) {//上一次出现的点数为k
                    if (j != k) {
                        for (int t = 1; t <= rollMax[k-1]; ++t) {
                            dp[i][j][1] += dp[i - 1][k][t];
                            dp[i][j][1] %= divider;
                        }
                    }
                    else {
                        for (int t = 1; t < rollMax[k-1]; ++t) {
                            dp[i][j][t + 1] += dp[i - 1][k][t];
                            dp[i][j][t + 1] %= divider;
                        }
                    }
                }
            }
        }
        int sum = 0;
        for (int i = 1; i < 7; ++i) {
            for (int t = 1; t <= rollMax[i-1]; ++t) {
                sum = (sum + dp[n - 1][i][t]) % divider;
            }
        }
        return sum;

    }
};